src/Loader.php

```diff
@@ -479,6 +479,140 @@ public static function register_scripts() {
+		$json_i18n_filenames       = glob( $lang_dir . $domain . '-' . $locale . '-*.json' );
+		$combined_translation_data = array();
+
+		foreach ( $json_i18n_filenames as $json_filename ) {
+			if ( ! $wp_filesystem->is_readable( $json_filename ) ) {
+				continue;
+			}
+
+			$file_contents = $wp_filesystem->get_contents( $json_filename );
+			$chunk_data    = \json_decode( $file_contents, true );
+
+			if ( empty( $chunk_data ) ) {
+				continue;
+			}
+
+			if ( ! isset( $chunk_data['comment']['reference'] ) ) {
+				continue;
+			}
+
+			$reference_file = $chunk_data['comment']['reference'];
+
+			// Only combine "app" files (not scripts registered with WP).
+			if (
+				false === strpos( $reference_file, 'dist/chunks/' ) &&
+				false === strpos( $reference_file, 'dist/app/index.js' )
+			) {
+				continue;
+			}
+
+			if ( empty( $combined_translation_data ) ) {
+				// Use the first translation file as the base structure.
+				$combined_translation_data = $chunk_data;
+			} else {
+				// Combine all messages from all chunk files.
+				$combined_translation_data['locale_data']['messages'] = array_merge(
+					$combined_translation_data['locale_data']['messages'],
+					$chunk_data['locale_data']['messages']
+				);
+			}
+		}
+
+		// Remove inaccurate reference comment.
+		unset( $combined_translation_data['comment'] );
+
+		return $combined_translation_data;
+	}
+
+	/**
+	 * Load translation strings from language packs for dynamic imports.
+	 *
+	 * This function combines JSON translation data auto-extracted by GlotPress
+	 * from Webpack-generated JS chunks into a single file that can be used in
+	 * subsequent requests. This is necessary since the JS chunks are not known
+	 * to WordPress via wp_register_script() and wp_set_script_translations().
+	 *
+	 * @param string $original_translations JSON encoded translations object.
+	 * @param string $file File location for the script being translated.
+	 * @param string $handle Script handle.
+	 * @param string $domain Text domain.
+	 *
+	 * @return string JSON encoded translations object.
+	 */
+	public static function load_script_translations( $original_translations, $file, $handle, $domain ) {
+		// Make sure the main app script is being loaded.
+		if ( WC_ADMIN_APP !== $handle ) {
+			return $original_translations;
+		}
+
+		// Make sure we're handing the correct domain (could be woocommerce or woocommerce-admin).
+		$plugin_domain = explode( '/', plugin_basename( __FILE__ ) )[0];
+		if ( $plugin_domain !== $domain ) {
+			return $original_translations;
+		}
+
+		$locale         = determine_locale();
+		$cache_filename = self::get_combined_translation_filename( $domain, $locale );
+		$lang_dir       = WP_LANG_DIR . '/plugins/';
+
+		// Allow us to easily interact with the filesystem.
+		require_once ABSPATH . 'wp-admin/includes/file.php';
+		\WP_Filesystem();
+		global $wp_filesystem;
+
+		// First attempt to get a previously generated combined translation file.
+		if (
```

Comment:
Had one more thought since our conversation on this topic yesterday. Regarding invalidation of the combined file I was thinking of the following scenario:

- This code ships in 4.4, translations are fixed and all is well in the world.
- We work on fab new features for 4.5, which include new strings. Translations happen for new strings
- After updating my site to 4.5, the old cache file in the lang directory still exists, so the code here returns the old file ( at least that is what i'm wondering could possibly happen ).

I didn't test that flow yet but was wondering if it _could_ happen. If so, wondering if we either need to add something to the update routine in each release to clear out the cached/combined translation file... or maybe we can add in the woocommerce_admin_version number to the file name to act as a cache buster in this scenario.

Reply:
There is a version string in the filename: https://github.com/woocommerce/woocommerce-admin/pull/4937/files#diff-3864ffe37f3a1de39876ea62cae525f3R491

Reply:
i missed that, thanks!!!